Friendships v2 feature mutual friendship code review request.

Byabasaija opened the pull request:
In this milestone feature, we added a mutual friendship functionality where
- A user can create a friendship with another user and automatically the friendship becomes bi-directional.

What we have implemented in this pull request:
-

**We have created a `FriendRequests`  model with**
- belongs to relationship with the request sender and receiver connected to the User model.

- 
**We have updated a `Friendship`  model with**
- belongs to relationship with user and friend
- Added methods to create and destroy inverse friendships

**We have added methods in `User` model**
 - `#confirm_friend(user)` to confirm the friendship request of user instance with the user on the argument passed.
 - `#friend_request?` to check if the passed user argument is the sender or receiver of the request.
 - `#friends?(user)` to check if the friendship between the user with the user provided in the argument.


**We have also added the unfriend feature.**
- It destroys a friendship belonging to the current user and this happens in a bi-directional way.

**NOTE:** In the integration tests we used rails fixtures to create the data we use in the test cases rather than creating the data directly from the spec files.
Any feedback given will be appreciated and highly considered :open_hands:

Byabasaija committed:
Add friend request controller modify accept and reject functionality

dipbazz committed:
Update user_spec.rb file to match the new changes

Remove unused test case for deleted method and update
existing methods to check the test for new criteria.

Byabasaija committed:
Implement destroy usefunctionality

Byabasaija committed:
Fix code with rubocop linters

mahmoud717 reviewed (changes requested):
### Changes required ♻️ 

**Hello @Byabasaija  & @dipbazz**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [x] in the friendship page you should display the actions buttons for users other than yourself and you should display them only for the user who is receiving the invitation when he receives an invitation.  right now it's displayed at all times for all users which is causing errors when clicked on. 

- [x] in the users' index page you should display pending if the request is sent and there is no response yet. as you should display friend if the user accepts the invitation.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

dipbazz committed:
Add user_helpers to check the friendships status

- Update user index view to dispaly the list of friend requests
  at top and all users at bottom.
- heper method to dispaly if request pending or is friend
- Update UI of list of users with bootstrap card layout.
- Remove friend_requests view which is not required.

enaburekhan reviewed (changes requested):
# Status: Request changes:full_moon:
**Hi team!
Great job in making all the changes requested by the previous Tse:clap::clap:. However, we just need to make a few changes to get this project approved:muscle:.**

**Mandatory requirement**

- [x] Check if students added Unit and Integration tests! ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

Currently, the test only covers some parts of the user model. We also need to test major functions in the controller. Both unit and integration test. Kindly fix this.

Make the above-required change and when you are done, request for another review.
You can drop your questions :point_down:if you have any and ping me on slack @ Eric Enaburekhan to respond.
Happy Coding:blush:
Eric Enaburekhan

Byabasaija committed:
Add controller_specs

Add user fixture
Add unit tests for controller index and show actions.

dipbazz committed:
Add integrations test case for freind request controller

Add fixture for friend requests to accept and reject the
friend request and create method to successfully test the
friend request sent and also check for redirect when the
user is not logged in.

Byabasaija committed:
Add tests for Friendships controller.

Create a fixture for friendships.

Temmarie reviewed (approved):
## Status: Approved :heavy_check_mark:

### Hi @Byabasaija and @dipbazz  ! :wave:

#### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: :tada:

 #### Optional
- [ ] Add unit tests for friendship.

#### You may merge, good luck with your next projects! :rocket:
#### Happy Coding! :+1:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)
